sdk/dapp-kit/src/components/AutoConnectWallet.tsx

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { ReactNode } from 'react';
+import { useEffect, useState } from 'react';
+import { useConnectWallet } from '../hooks/wallet/useConnectWallet.js';
+import { useWalletContext } from '../hooks/wallet/useWalletContext.js';
+import type { StorageAdapter } from '../utils/storageAdapters.js';
+import type { WalletAccount, WalletWithRequiredFeatures } from '@mysten/wallet-standard';
+
+type AutoConnectWalletProps = {
+	storageKey: string;
+	storageAdapter: StorageAdapter;
+	children: ReactNode;
+};
+
+export function AutoConnectWallet({
+	storageKey,
+	storageAdapter,
+	children,
+}: AutoConnectWalletProps) {
+	const { wallets, currentWallet: wallet, currentAccount: account } = useWalletContext();
+	const [previousWallet, setPreviousWallet] = useState<WalletWithRequiredFeatures | null>(null);
+	const [currentWallet, setCurrentWallet] = useState<WalletWithRequiredFeatures | null>(wallet);
+	const [previousAccount, setPreviousAccount] = useState<WalletAccount | null>(null);
+	const [currentAccount, setCurrentAccount] = useState<WalletAccount | null>(account);
+	const { mutate: connectWallet } = useConnectWallet();
+
+	// Instead of abstracting the previous wallet and account state logic into a generic
+	// usePrevious hook, we'll write this plainly to make it clear that this code depends
+	// on triggering shallow re-renders by updating state in the render code.
+	if (wallet !== currentWallet) {
+		setPreviousWallet(currentWallet);
+		setCurrentWallet(wallet);
+	}
+
+	if (account !== currentAccount) {
+		setPreviousAccount(currentAccount);
+		setCurrentAccount(account);
+	}
+
+	useEffect(() => {
+		if (wallet !== previousWallet || account !== previousAccount) {
+			if (wallet) {
+				setWalletConnectionInfo({
+					storageAdapter,
+					storageKey,
+					walletName: wallet.name,
+					accountAddress: account?.address,
+				});
+			} else {
+				removeWalletConnectionInfo(storageAdapter, storageKey);
+			}
+		}
+	}, [account, previousAccount, previousWallet, storageAdapter, storageKey, wallet]);
```

Comment:
will this effect be run twice anytime the wallet/account changes (once for the change, once for the previousX change)?  The timing of when the various pieces of state get updated here seem a little harry, and might benefit from some comments.

Reply:
I think there are extra effect calls here, which is one thing that makes this approach not-so-great from a theoretical perf standpoint (see comment above)